Change cumulative compaction for decoupling storage from compution

chaoyli opened the pull request:
1. Calculate cumulative point when loading tablet first time.
2. Simplify pick rowsets logic upon delete predicate.
3. Saving meta and modify rowsets only once after cumulative compaction.

Commit:
Change cumulative compaction for decoupling storage from compution

1. Calculate cumulative point when loading tablet first time.
2. Simplify pick rowsets logic upon delete predicate.
3. Saving meta and modify rowsets only once after cumulative compaction.

Commit:
Change base compaction for decoupling storage from compution

1. Simplify pick rowsets logic.
2. Saving meta and modify rowsets only once after base compaction.

imay commented on be/src/olap/base_compaction.h:
```diff
@@ -32,152 +33,45 @@ namespace doris {
-    // - 如果执行成功，则返回OLAP_SUCCESS；
-    // - 其它情况下，返回相应的错误码
-    OLAPStatus run();
+    BaseCompaction(TabletSharedPtr tablet);
```
you should add comments for these public functions. It's better to add an example for how to use this class.

If some functions aren't intended to be used by others, please make them private

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
 
 namespace doris {
```
It's better to move the rowset_comparator to rowset.h file

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
-        return OLAP_ERR_BE_TRY_BE_LOCK_ERROR;
+BaseCompaction::~BaseCompaction() {
+    if (_base_locked) {
+        _tablet->release_base_compaction_lock();
```
It's too trick. Add a scoped lock or mutex.

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
-        return OLAP_ERR_BE_TRY_BE_LOCK_ERROR;
+BaseCompaction::~BaseCompaction() {
+    if (_base_locked) {
+        _tablet->release_base_compaction_lock();
```
It's too trick to add a variable _base_locked to indicate whether base compaction lock is owned. It's better to add a scoped mutex like migration lock.

chaoyli replied:
OK, I will change it to another form

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
-        DorisMetrics::base_compaction_bytes_total.increment(merge_bytes);
+OLAPStatus BaseCompaction::pick_rowsets_to_compact() {
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
```
should use capture rs readers to check data is valid

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
-    WriteLock wrlock(_tablet->get_header_lock_ptr());
+OLAPStatus BaseCompaction::modify_rowsets() {
+    std::vector<RowsetSharedPtr> output_rowsets;
+    output_rowsets.push_back(_output_rowset);
```
Need tablet write lock？

yiguolei commented on be/src/olap/cumulative_compaction.cpp:
```diff
@@ -32,356 +32,121 @@ using std::vector;
-        _tablet->release_cumulative_lock();
-        return OLAP_ERR_CUMULATIVE_INVALID_PARAMETERS;
-    }
+    _cumulative_locked = true;
```
Use scoped lock here

yiguolei commented on be/src/olap/cumulative_compaction.cpp:
```diff
@@ -32,356 +32,121 @@ using std::vector;
+    if (_input_rowsets.empty()) {
+        _tablet->set_cumulative_layer_point(candidate_rowsets.back()->end_version() + 1);
+    } else {
+        _tablet->set_cumulative_layer_point(_input_rowsets.back()->end_version() + 1);
```
Must not update ce point here. Should update it after ce success or ce failed

yiguolei commented on be/src/olap/tablet.cpp:
```diff
@@ -736,6 +749,32 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
     return OLAP_SUCCESS;
 }
 
+OLAPStatus Tablet::calcuate_cumulative_point() {
```
it is not used out of tablet. set it to private

yiguolei commented on be/src/olap/tablet.cpp:
```diff
@@ -736,6 +749,32 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
 
+OLAPStatus Tablet::calcuate_cumulative_point() {
+    ReadLock rdlock(&_meta_lock);
+    std::list<Version> existing_versions;
```
It is used during init. Does it need meta lock really?

yiguolei commented on be/src/olap/tablet.cpp:
```diff
@@ -933,4 +972,20 @@ OLAPStatus Tablet::set_partition_id(int64_t partition_id) {
+void Tablet::pick_candicate_rowsets_to_cumulative_compaction(std::vector<RowsetSharedPtr>* candidate_rowsets) {
+    for (auto& it : _rs_version_map) {
+        if (it.first.first > _cumulative_point) {
+            candidate_rowsets->push_back(it.second);
```
add first == second here

yiguolei commented on be/src/olap/base_compaction.cpp:
```diff
@@ -40,487 +39,264 @@ using std::vector;
-
-    LOG(INFO) << "init base compaction handler. [tablet=" << tablet->full_name() << "]";
+static bool rowset_comparator(const RowsetSharedPtr& left, const RowsetSharedPtr& right) {
+    return left->end_version() < right->end_version();
```
Should add a parent class and move all common methods and methods only used by compaction to parent class. There are too many duplicate methods.

yiguolei reviewed (commented):
some comments

Commit:
Extract Compaction Class from CumulativeCompaction and BaseComapction.

1. Add annotation to Compaction Class
2. Fix set cumulative_layer_point bug before cumulative compaction completion.

yiguolei commented on be/src/olap/tablet.cpp:
```diff
@@ -736,6 +749,31 @@ OLAPStatus Tablet::max_continuous_version_from_begining(Version* version, Versio
+
+    _cumulative_point = -1;
+    int64_t prev_version = -1;
+    for (const Version& version : existing_versions) {
```
prev_version not updated?

Commit:
Fix calculate cumulative point bug

yiguolei reviewed (dismissed):
LGTM

imay commented on be/src/olap/tablet_manager.cpp:
```diff
@@ -737,18 +737,16 @@ TabletSharedPtr TabletManager::find_best_tablet_to_compaction(
             if (compaction_type == CompactionType::BASE_COMPACTION) {
-                if (!table_ptr->try_base_compaction_lock()) {
+                MutexLock lock(table_ptr->get_base_lock());
+                if (!lock.own_lock()) {
```
This is not a try-lock, what does this check mean?

chaoyli replied:
I missed it. I will add it in next commit.

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,511 +16,145 @@
+OLAPStatus BaseCompaction::pick_rowsets_to_compact() {
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
+    if (_input_rowsets.size() == 0 || _input_rowsets.size() == 1) {
```
```suggestion
    if (_input_rowsets.size() <= 1) {
```

imay commented on be/src/olap/tablet.cpp:
```diff
@@ -933,4 +973,21 @@ OLAPStatus Tablet::set_partition_id(int64_t partition_id) {
+}
+
+void Tablet::pick_candicate_rowsets_to_base_compaction(std::vector<RowsetSharedPtr>* candidate_rowsets) {
+    for (auto& it : _rs_version_map) {
```
Is this thread-safe?

chaoyli replied:
I will add read lock to it.

imay commented on be/src/olap/storage_engine.cpp:
```diff
@@ -598,27 +585,13 @@ void StorageEngine::perform_base_compaction(DataDir* data_dir) {
     TabletSharedPtr best_tablet = _tablet_manager->find_best_tablet_to_compaction(CompactionType::BASE_COMPACTION, data_dir);
     if (best_tablet == nullptr) { return; }
 
-    DorisMetrics::base_compaction_request_total.increment(1);
```
why do you drop this metrics?

chaoyli replied:
I drop it wrong. I add it back.

xy720 committed:
Merge Help document to documentation (#1586)

Help document collation (integration of help and documentation documents)

Commit:
Add mutex to thread-safe

yiguolei commented on be/src/olap/tablet_manager.cpp:
```diff
@@ -301,6 +301,11 @@ OLAPStatus TabletManager::create_tablet(const TCreateTabletReq& request,
+    // If not, cumulative_point will set to minus one
+    // when init tablet during create tablet.
+    RETURN_NOT_OK(tablet->calculate_cumulative_point());
+
```
If calculate method is called in two cases, it's better to call it during cumulative compaction.

yiguolei reviewed (commented):
some comments

Commit:
Move calculate_cumulative_point() logic to cumulative compaction

yiguolei reviewed (dismissed):
LGTM

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,511 +16,151 @@
+    _input_rowsets.clear();
+    _tablet->pick_candicate_rowsets_to_base_compaction(&_input_rowsets);
+    if (_input_rowsets.size() <= 1) {
+        LOG(INFO) << "There is no enough rowsets to cumulative compaction."
```
Do we need this here as INFO?

chaoyli replied:
I will change it to LOG(WARNING)

imay replied:
I think we should remove this LOG.

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,511 +16,151 @@
+        _input_rowsets_size += rowset->data_disk_size();
+        _input_row_num += rowset->num_rows();
     }
+    DorisMetrics::base_compaction_bytes_total.increment(_input_rowsets_size);
```
I think we should update metric when compaction is finished

And this metric can be get from compaction process

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,511 +16,151 @@
-    uint64_t merged_rows = 0;
-    uint64_t filted_rows = 0;
-    OLAPStatus res = OLAP_SUCCESS;
+    RETURN_NOT_OK(construct_output_rowset_writer());
```
I think it's better to encapsulate all the following logic into base class.

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,511 +16,151 @@
-            return OLAP_ERR_ROWSET_CREATE_READER;
-        }
-        rs_readers.push_back(rs_reader);
+OLAPStatus BaseCompaction::do_compaction() {
```
I think it's better to encapsulate all the following logic into base class.

chaoyli replied:
I do not think so.
Different Compaction can has different logic. 
If I want to change base compaction to capture partial rowsets, I can do it in this function.

imay replied:
most of this function's logic is the same with cumulative compaction. 
1. pick rowset
2. check rowset if continue
3. create readers and writer
4. check rows if match
5. write metadata

Now, there are two same logic in two files, I think we can make it better.

Commit:
Change log level

Commit:
Extract common code to base class

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,513 +16,115 @@
-
-    return OLAP_SUCCESS;
+ReaderType BaseCompaction::compaction_type() const {
+    return _compaction_type;
```
there is no need to store it in class attribute, It's OK to just return a constant value.

imay commented on be/src/olap/compaction.h:
```diff
@@ -0,0 +1,89 @@
+public:
+    Compaction(TabletSharedPtr tablet);
+    virtual ~Compaction();
+
```
Only keep needed function as public, make others private/protected

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,513 +16,115 @@
-    VLOG(10) << "valid need merged version";
-    return true;
+const std::string& BaseCompaction::compaction_name() const {
+    return _compaction_name;
```
Just return const value too

chaoyli replied:
I want to use reference to reduce string copy.
If use const value, it will cause string copy.

Commit:
Make function only used in compaction protected

yiguolei reviewed (commented):
LGTM

imay commented on be/src/olap/base_compaction.cpp:
```diff
@@ -16,513 +16,114 @@
-    VLOG(3) << "delete file action is OK";
-
-    return OLAP_SUCCESS;
+ReaderType BaseCompaction::compaction_type() const {
```
put this to .h file

imay commented on be/src/olap/base_compaction.h:
```diff
@@ -18,166 +18,30 @@
-    std::vector<RowsetSharedPtr> _new_rowsets;
-    bool _base_compaction_locked;
-    RowsetWriterSharedPtr _rs_writer;
+    std::string _compaction_name;
```
delete these

Commit:
Move codes in cppl file to header file

imay reviewed (approved):
LGTM
